RFC: Encourage consistent usage of api helper

bkjohnson opened the pull request:
## Description

This is part of https://github.com/department-of-veterans-affairs/va.gov-team/issues/224

The `apiRequest` helper currently allows itself to be used in many different ways, and this has lead to inconsistencies in the code for veteran-facing apps.  With this PR I am suggesting a pattern that would encourage a much stricter (and therefore more consistent) usage of our fetch wrapper.

### Current examples

#### Callback arguments
The request helper allows a `success` and a `failure` callback` to be passed as arguments as shown here:
https://github.com/department-of-veterans-affairs/vets-website/blob/f4a6e90875c5ab503e1933bdb6134e32958cea69/src/platform/user/profile/actions/mhv.js#L37-L45

#### Promise chaining
Alternatively, you can use promise chaining instead of callbacks:
https://github.com/department-of-veterans-affairs/vets-website/blob/6a27450abed0eb724d9fa0334633ff8ffe700f78/src/applications/beta-enrollment/actions/index.js#L17-L19

Each of these examples also demonstrates the `method` to be used and the `url` to be fetched, both of which can either be inlined or passed as a variable


### Proposal


### :warning: The following proposal is no longer relevant - see [this comment](https://github.com/department-of-veterans-affairs/vets-website/pull/10845#issuecomment-540692196) for the current proposal. :warning: 

For a concrete example I will use the url from the previous example: `/beta_registration/${service}`.

The pattern I'm proposing would have us fetch that url like this:

```js
api.post.beta_registrations[service]()
  .then(() => dispatch({ type: REGISTER_SERVICE, service }))

```

This works by using a [JS Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy).  It is set up so that the first property accessed specifies _what_ will be done to the resource (the method), and any property after that is treated as a chunk in the resource url.  The call is invoked with `()`, and any optional body params or query string params can be passed in as arguments.

What are the overall thoughts to this approach?  The idea is to make what is happening behind each request a bit more obvious by specifying the method upfront and relying on promise chaining rather than callback arguments.

bkjohnson committed:
Initial demo of Proxy-based api request helper

bkjohnson committed:
Add more tests for GET requests

bkjohnson committed:
Add tests for GET requests w/ query params

bkjohnson committed:
Add test for POST requests

bkjohnson committed:
Add test for DELETE requests

bkjohnson committed:
Add test for PATCH request

bkjohnson committed:
Add test for unsupported HTTP method

bkjohnson committed:
Add tests for constructUrl helper function

bkjohnson committed:
Fix failing tests due to fetch not being mocked

rianfowler commented on src/platform/utilities/api/service/index.js:
```diff
@@ -0,0 +1,103 @@
+  get(obj, prop) {
+    this.chunks.push(prop); // keep accumulating the path chunks
+
+    return new Proxy(sendFetch.bind({ ...this }), resourceHandler);
```
never used a Proxy before so LMK if this correct understanding: 

```
const service = 'somePathPartString`;

api.post.beta_registrations[service]()
```

In this example, `resourceHandler.get` is called twice- once with `prop = beta_registrations` and again with `prop = service` and on each call `sendFetch` is bound to a new object that's a copy of the current `resourceHandler`. 

why is `sendFetch` bound to a copy of `resourceHandler` and not just `this`? Is it because `resourceHandler` is shared and binding it that way creates a state for a specific api request instance?

erikphansen replied:
Can we even use Proxies in IE11?

bkjohnson replied:
Yeah, JS Proxies are a bit different than traditional programming so wrapping your head around them can be tricky at first.

`sendFetch` isn't bound to a copy of `resourceHandler` - it is bound to a copy of `this`.  The [proxy constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy#Syntax) looks like this:

```js
var p = new Proxy(target, handler);
```

The `resourceHandler` has a `get()` function that acts as a trap for any property-accessing that happens, and that trap has this in it:

```js
return new Proxy(sendFetch.bind({ ...this }), resourceHandler);
```

Returning a Proxy with a recursive `resourceHandler` as the handler allows us to chain as many resource chunks as we want on to a request ( `api.get.whatever.you.want.forever()`).

The state for each request gets stored in `this`.  Every time `resourceHandler` has a property accessed, it appends that property to the `this.chunks` array which is then used to construct the url.

rianfowler replied:
I think erik is right- there isn't a polyfill for Proxy that works with IE 11. :(

erikphansen replied:
I think it was my first week on the job that someone broke production because we used Proxies :-)

bkjohnson replied:
That's a bummer.  No IE [support](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy#Browser_compatibility) but they do work on Edge.  Do we know how much longer we'll have to support IE?

cvalarida replied:
Indefinitely, as far I know. I vaguely remember hearing once upon a time that there are a lot of VA facilities that are running IE on older machines for Veterans to use.

I think our official browser support has something to do with the popularity of the browser, but I don't recall if there's an exception for IE because of those VA facilities.

rianfowler commented on src/platform/utilities/api/service/index.js:
```diff
@@ -0,0 +1,103 @@
+  let url = `${host}/${chunks.join('/')}`;
+  if (typeof arg === 'number') {
+    url += `/${arg}`;
+  } else if (arg != null && typeof arg === 'object' && method === METHODS.get) {
```
can you show us an example of how arg is used? is it like 
```
api
.post
.beta_registrations[service]({ query1: "value", query2: "value" })
```

bkjohnson replied:
With the current setup, arg is used differently for a GET and a POST.  For a GET the arg is encoded as a query string like this:

```js
// methodHandler receives a `get`, so the arg becomes a query string
api.get.appointments({ foo: 'bar', hello: 'world' })
```
https://github.com/department-of-veterans-affairs/vets-website/blob/3ccb87a541ee670131fa3263fcf4f68132cb53ad/src/platform/utilities/api/service/api.unit.spec.js#L50-L71

</br>
For a POST, the arg is treated as json that needs to be stringified and put in the request body.  (We could also have it handle multipart form data if we wanted to).  Here are some examples of that:

The `post` here would result in the invocation treating the same arg as a body object rather than a query string:

```js
// methodHandler receives a `post`, so the arg becomes the body content
api.post.appointments({ foo: 'bar', hello: 'world' })
```

https://github.com/department-of-veterans-affairs/vets-website/blob/3ccb87a541ee670131fa3263fcf4f68132cb53ad/src/platform/utilities/api/service/api.unit.spec.js#L74-L84

rianfowler reviewed (commented):
I think this looks good. I would like to see some more examples of how to make GETs and POSTs and what the output / behavior of each is (e.g. query strings, posts bodies).

bkjohnson commented:
Since proxies are out of the question, what about removing `success` and `error` callback args and also removing this redirect behavior and throwing an error instead:

https://github.com/department-of-veterans-affairs/vets-website/blob/28eb457d9523f0e4d743eba15648c21a44ef553d/src/platform/utilities/api/index.js#L85-L95

rianfowler commented:
I think @U-DON may need to comment on removing the redirect behavior from the helper- i'm not really sure what it's for. 

i'm fine with removing the success / error callbacks in the interest of reducing of the number of ways it can be used.

U-DON commented:
I'm not sure we should remove the redirect logic. I don't necessarily like it in the `apiRequest` function, but I don't think any app currently handles a 401 error to properly show the session expired page. However, I know there was a session timeout feature that would eliminate the need for this redirect. That feature is still currently gated on production, though.

https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/site-wide/user-nav/containers/Main.jsx#L167-L172

rianfowler commented:
@U-DON 
- Who is working on the session timeout modal? 
- Do you think it covers all of the use cases covered by the api request helper?

---
I think it would be good to make people reference the redirect behavior explicitly (e.g. function you pass explicitly into the promise chain) or maybe rename the function. I think there are a few services where we wouldn't want to redirect (e.g. facility locator, gibct). 

As a maintainer, I think it would be nice to see all of the errors that are handled in single place (without having to read a lot of code). We could establish a pattern using the promise chain e.g. 

```
const handleSuccess = response => ... ; // handle success

import { handleNot200, redirectOn404) from 'platform/.../fetchErrorHandlers'; 

return apiRequest(`/service`, settings) 
.then(redirectOn404) // or maybeRedirectOn404 
.then(handleNot200) 
.then(handleSuccess) 
...
```

We could suggest a way to organize this logic (e.g. "you should use a switch statement to handle your errors") or we could provide a way that makes it easy to add straight forward error handling. Might be a better way to do this with `async` `await`, too.

bkjohnson commented:
I like the plugin-style handlers to cover certain specific cases.  However, if we go that route, we have to consider what the main helper itself is returning.  Currently we have this:

```js
    const data = isJson(response)
      ? response.json()
      : Promise.resolve(response);
  
    if (response.ok || response.status === 304) {
      return data;
    }
```

So sometimes the `apiRequest` promise would resolve to json, and other times it would just resolve as the response.  In order to support those pre-defined handlers for the promise chain we will need to always return the full response, correct?

rianfowler commented:
Yeah- we could just get rid of the helper and only provide response handlers. I'm not sure, though- I haven't ever seen this pattern used. Mostly just trying to imagine how we could make it easier for maintainers by making it more explicit while also not adding too much boilerplate.

U-DON commented:
@rianfowler: I think the session timeout modal is pretty much complete but just disabled for production. @narin was the last person to work on it.

It doesn't really intersect with `apiRequest` aside from the behavior of redirecting when a session times out. The benefit of enabling the timeout modal is that the `apiRequest` helper won't have to concern itself with the 401 response.

--

While we're discussing the finer points of abstracting HTTP requests, I started thinking maybe we can use a third party library. [`axios`](https://github.com/axios/axios) came to mind. It seems simple to use and has some of these advantages I gleaned from a first glance:
- Automatically parses JSON responses.
- Built in XSRF protection.
- Provides shorthands like `axios.post(url, params)`.
- Can add interceptors for requests and responses, which are functions that run before requests and responses. The 401 handling is one that comes to mind if we decide to keep it.
- Loading progress for downloads and uploads if we ever want to make use of it.

narin commented:
Hey all.  I worked on the session timeout modal last and am pretty sure the work on it is 95% complete and may just need to have the `isProduction()` flag removed and tested.  I'm not sure who owns that body of work now, as I was switched over to the VAOS project.  I'd be happy to assist with this, whether it be wrapping up the work or helping out whoever owns it now.

In regards to `apiRequest()`, I _did_ make some modifications to this to support the session timeout modal.  In order to keep the session expiration datetime up to date on the front end, all calls to `vets-api` return a `X-Session-Expiration` header.  You'll see that I wrote the custom fetch function called `fetchAndUpdateSessionExpiration()` in `src/platform/utilities/api/index.js` which `apiRequest()` uses.  It works exactly the same way `fetch()` does but it also checks for this header and updates the session expiration stored in localStorage.

The reasoning for this is because not all API calls in the application are consistently using `apiRequest()`.  Some were using a plain `fetch()` instead.  Modifying all of the calls to use `apiRequest()` would have require quite a bit of refactoring, especially since all of them handle errors differently, so I wrote the `fetchAndUpdateSessionExpiration()` function, and replaced all instances of `fetch()` in the app with this custom function.  This solution wasn't ideal but I was constrained by time right before the previous contract ended. Apologies if it further complicates things.

Hope this clears things up a bit!

bkjohnson committed:
Remove PRoxy-based api helper

We won't be going in this direction as Proxies aren't supported in IE11

bkjohnson commented:
I've been thinking more about the callback arguments and different ways to approach that.  A safe option might be to add something like this towards the top of the helper (and an additional check for the `error` callback):

```js
   if (success) {
     // eslint-disable-next-line no-console
     console.warn(
       'the "success" callback has been deprecated, please use a promise chain instead',
     );
  } 
```

At the very least this would make test output a bit uglier which might encourage VFS developers to make the switch towards promise chaining.

A more aggressive approach would be to modify the helper so that it doesn't accept `success` and `error` as arguments and also refactor all of the code to move those arguments into the promise chain.  I'm tempted to see what that would look like because in addition to the gain from moving things into a promise chain, this would avoid the pattern of passing in `null` for the options just as a placeholder.

bkjohnson committed:
Refactor platform code to use promise chaining

bkjohnson committed:
Refactor search app

bkjohnson committed:
Refactor vic-v2 app

bkjohnson committed:
Refactor vic-v2 app, pt. 2

I meant to include this change with the previous commit

bkjohnson committed:
Refactor letters app

bkjohnson committed:
Refactor static-pages app

bkjohnson committed:
Refactor burials app

bkjohnson committed:
Refactor hca app

bkjohnson committed:
Refactor veteran-id-card app

bkjohnson committed:
Refactor auth app

bkjohnson committed:
Refactor personalization app

bkjohnson committed:
Refactor id-card-beta app

bkjohnson committed:
Refactor edu-benefits app

bkjohnson committed:
Refactor terms-and-conditions app

bkjohnson committed:
Refactor disability-benefits app

bkjohnson committed:
Log warning msg when success/error callback used

Hopefully this can be a temporary change to encourage developers to
rafactor their code to use promise chaining.  Once that happens the
success and error args can be removed from the function entirely.

bkjohnson commented:
There are two apps remaining that I haven't refactored to use promise chaining:

**1. `claims-status`**

The `apiRequest` helper is used with the callback args:
https://github.com/department-of-veterans-affairs/vets-website/blob/fa7f37f7d6df96d68f374de77ae139c947ab3881/src/applications/claims-status/actions/index.jsx#L140-L175

And there are tests which rely on spying on the helper which would currently break if the code was refactored as a promise chain:

https://github.com/department-of-veterans-affairs/vets-website/blob/fa7f37f7d6df96d68f374de77ae139c947ab3881/src/applications/claims-status/tests/actions.unit.spec.js#L426

https://github.com/department-of-veterans-affairs/vets-website/blob/fa7f37f7d6df96d68f374de77ae139c947ab3881/src/applications/claims-status/tests/actions.unit.spec.js#L440-L445


**2. `post-911-gib-status`**

The helper is used in a way where the `success` and `error` callbacks are passed as args, but there is also a `.catch`.  This suggests that the final catch is meant to handle any error thrown in the `error` callback, but I'm not sure if this design is intentional or not and I would rather leave it as is to be safe.

https://github.com/department-of-veterans-affairs/vets-website/blob/fa7f37f7d6df96d68f374de77ae139c947ab3881/src/applications/post-911-gib-status/actions/post-911-gib-status.js#L19-L63


How do people feel about what I've done:
- refactoring the easy parts
- [adding the warning messages](https://github.com/department-of-veterans-affairs/vets-website/pull/10845/commits/1fa876807bdc461dcdd298d32765a4973ffdc94c) to encourage developers to refactor their code to the new pattern.

Should we have a plan to keep an eye on this so that we can remove the `success` and `error` callbacks once the developers have refactored their code to not use them?  I'd like to fully remove them once we know it won't break anything.

bkjohnson committed:
Merge branch 'master' into va-team-224-api-helper

rianfowler reviewed (approved):
This looks good to me. 

- long term we may want to find a library to handle polling. the claims status was our attempt at making something generic that would be easy to update / understand. I think the idea was to make it easy to unit test predicates + handlers without requiring people to write a bunch of test logic to work around the polling logic.
-  not sure about the post-911 handlers- my understanding is that you can only `catch` network errors (e.g. not connected to the internet) with `fetch`. might be something useful to look into or train on later (e.g. if there are other scenarios that `catch` works for and if we should standardize handling that in our ux / error reporting)
